stacks-signer/src/stacks_client.rs

```diff
@@ -98,8 +132,122 @@ impl StacksClient {
+                payer_spending_condition.set_tx_fee(tx_fee);
+                TransactionAuth::Sponsored(sender_spending_condition, payer_spending_condition)
+            }
+            _ => {
+                sender_spending_condition.set_tx_fee(tx_fee);
+                TransactionAuth::Standard(sender_spending_condition)
+            }
+        };
+        let mut unsigned_tx = StacksTransaction::new(version, auth, payload);
+        unsigned_tx.anchor_mode = anchor_mode;
+        unsigned_tx.post_condition_mode = TransactionPostConditionMode::Allow;
+        unsigned_tx.chain_id = chain_id;
+
+        let mut tx_signer = StacksTransactionSigner::new(&unsigned_tx);
+        tx_signer.sign_origin(sender).unwrap();
+        if let (Some(payer), Some(_)) = (payer, payer_nonce) {
+            tx_signer.sign_sponsor(payer).unwrap();
+        }
+
+        let mut buf = vec![];
+        tx_signer
+            .get_tx()
+            .unwrap()
+            .consensus_serialize(&mut buf)
+            .unwrap();
+        buf
+    }
 }
 
+impl InteractWithStacksContracts for StacksClient {
+    fn make_contract_call(
+        sender: &StacksPrivateKey,
+        nonce: u64,
+        contract_addr: &StacksAddress,
+        contract_name: &str,
+        function_name: &str,
+        function_args: &[ClarityValue],
+        tx_version: TransactionVersion,
+        chain_id: u32
+    ) -> Vec<u8> {
+        let contract_name = ContractName::from(contract_name);
+        let function_name = ClarityName::from(function_name);
+
+        let payload = TransactionContractCall {
+            address: contract_addr.clone(),
+            contract_name,
+            function_name,
+            function_args: function_args.iter().map(|x| x.clone()).collect(),
+        };
+
+        let tx_fee = 0;
+
+        Self::seralize_sign_tx_anchor_mode_version(
+            payload.into(),
+            sender,
+            None,
+            nonce,
+            None,
+            tx_fee,
+            TransactionAnchorMode::OnChainOnly,
+            tx_version,
+            chain_id
+        )
+    }
+
+    fn submit_tx(http_origin: &str, tx: &Vec<u8>) -> String {
+        let client = reqwest::blocking::Client::new();
+        let path = format!("{}/v2/transactions", http_origin);
+        let res = client
+            .post(&path)
+            .header("Content-Type", "application/octet-stream")
+            .body(tx.clone())
+            .send()
+            .unwrap();
+        if res.status().is_success() {
+            let res: String = res.json().unwrap();
+            assert_eq!(
+                res,
+                StacksTransaction::consensus_deserialize(&mut &tx[..])
+                    .unwrap()
+                    .txid()
+                    .to_string()
+            );
+            return res;
+        } else {
+            eprintln!("Submit tx error: {}", res.text().unwrap());
+            panic!("");
```

Comment:
we will have to turn this probably into some sort of retry with backoff logic and never panic.

Reply:
Would you mind updating this to return a result. And then move the "FIXME" to the panic to the external calling of this function? Think its better for the traits to never panic and leave it to the caller to panic.

Reply:
Seconded.  Let's not panic if we can help it.  Panicking is the library consumer's job :)